### ISSUE TITLE
no-useless-escape rule is not handling regex back references correctly

### ISSUE THREAD

**Issue by jvswatson** [eslint/eslint#7472]
It appears that the recent changes related to the no-useless-escape rule are not handling the integer back references (e.g. `\1`) correctly.

eslint: 3.9.0
node: 4.4.3
npm: 2.15.1
parser: default

.eslintrc:

```
{
  "rules": {
    "no-useless-escape": 2,
    "no-console": 0
  }
}
```

JavaScript file:

``` js
var keyword = "'the thing \"on the inside\" is what to match'";
var exactMatchPhrases = keyword.match(/(["'])(.*?)\1/g);
console.log(exactMatchPhrases);
```

**What did you expect to happen?**
No eslint errors

**What actually happened? Please include the actual, raw output from ESLint.**
2:51  error  Unnecessary escape character: \1  no-useless-escape

**Comment by not-an-aardvark** [comment]
Confirmed, thanks for the report.

**Comment by not-an-aardvark** [comment]
I think the issue is that digits in regex literals should always be valid escapes (either they represent control characters, or they represent backreferences).

```js
var foo = /\0/; // null character
var foo = /\1/; // \x01 character
var foo = /(a)\1/; // 2 'a' characters in a row
```

**Comment by ljharb** [comment]
Another example: `/^(['"]).*\1$/`
